service-worker.js

```diff
@@ -339,7 +340,11 @@ function fetchRequestFromZIM(fetchEvent) {
  * @returns {String} The same URL without its parameters and anchors
  */
 function removeUrlParameters(url) {
-    return url.replace(/([^?#]+)[?#].*$/, '$1');
+    // Remove any querystring
+    var strippedUrl = url.replace(/\?[^?]*$/, '');
+    // Remove any anchor parameters
+    strippedUrl = strippedUrl.replace(/#[^#]*$/, '');
```

Comment:
There is a potential weakness here. A valid encoded URI might have for example: 'My_friend&#39;s_house'. This regexp would strip everything after the #. Note that running decodeURI or decodeURIComponent does **not** decode that as an apostrophe. In practice, I don't believe our ZIMs are using this kind of encoding. Working around it is tricky because negative lookbehinds are not supported in any browsers except recent Chromium, so we can't easily search for a # that is not preceded by an ampersand. We could use the trick of reversing the string and using a negative lookahead (which is supported in all browsers). The logic is a bit mind-bending, though.

Reply:
OK, although in HTML5 almost anything can be an ID for an element, in practice JavaScript limits this to the following for backward compatibility with HTML4:

> ID and NAME tokens must begin with a letter ([A-Za-z]) and may be followed by any number of letters, digits ([0-9]), hyphens ("-"), underscores ("_"), colons (":"), and periods (".").

Source: https://stackoverflow.com/a/79022/9727685

In fact, `document.getElementById` cannot select an id that begins with a number, and colons and periods have to be escaped to use CSS selectors for obvious reasons.

In practice it should be safe to say that if we find a semicolon after the # then it is **not** an anchor and we shouldn't strip it, as this will allow us to distinguish any ampersand-encoded characters that use #.

Reply:
Maybe a safer approach could be to let the browser do that for us.
I did not test, but we might create a https://developer.mozilla.org/en-US/docs/Web/API/URL object from that string, then reset its `search` and `hash` properties. Or maybe even read its `pathname` property?

Reply:
The issue is IE11 not supporting the properties of a URL that we would need to access. We could do something similar to what we did with `uiUtil.deriveZimUrlFromRelativeUrl()` , which is to use the API if it's available and fall back to a regex if not (assuming the API works for our use case).

Reply:
Here we could safely use it, as we're inside a serviceworker (that will not be executed in IE11).

Reply:
OK, true. We also have the same funciton in uiUtil, (unfortunately we have to repeat it) so ideally we should use the same code which will, as you say, use the API in SW mode universally, even though it will have the fallback in it (and we should have a note in both implementations to say that the code must be kept synchronous). I can have a go at this.

Reply:
I'm OK with that